src/Bridge/Doctrine/Orm/Util/QueryBuilderHelper.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Doctrine\Orm\Util;
+
+use Doctrine\ORM\Query\Expr\Join;
+use Doctrine\ORM\QueryBuilder;
+
+/**
+ * @author Vincent Chalamon <vincent@les-tilleuls.coop>
+ *
+ * @internal
+ */
+final class QueryBuilderHelper
+{
+    /**
+     * Adds a join to the queryBuilder if none exists.
+     *
+     * @param QueryBuilder                $queryBuilder
+     * @param QueryNameGeneratorInterface $queryNameGenerator
+     * @param string                      $alias
+     * @param string                      $association        the association field
+     * @param string|null                 $joinType           the join type (left join / inner join)
+     * @param string|null                 $conditionType
+     * @param string|null                 $condition
+     *
+     * @return string the new association alias
+     */
+    public static function addJoinOnce(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $alias, string $association, $joinType = null, $conditionType = null, $condition = null): string
+    {
+        $join = self::getExistingJoin($queryBuilder, $alias, $association);
+
+        if (null === $join) {
+            $associationAlias = $queryNameGenerator->generateJoinAlias($association);
+            $query = sprintf('%s.%s', $alias, $association);
+
+            if (Join::LEFT_JOIN === $joinType || true === QueryChecker::hasLeftJoin($queryBuilder)) {
+                $queryBuilder->leftJoin($query, $associationAlias, $conditionType, $condition);
+            } else {
+                $queryBuilder->innerJoin($query, $associationAlias, $conditionType, $condition);
+            }
+        } else {
+            $associationAlias = $join->getAlias();
+        }
+
+        return $associationAlias;
+    }
+
+    /**
+     * Get the existing join from queryBuilder DQL parts.
+     *
+     * @param QueryBuilder $queryBuilder
+     * @param string       $alias
+     * @param string       $association  the association field
+     *
+     * @return Join|null
+     */
+    private static function getExistingJoin(QueryBuilder $queryBuilder, string $alias, string $association)
+    {
+        $parts = $queryBuilder->getDQLPart('join');
+
+        if (!isset($parts['o'])) {
```

Comment:
Maybe we could put this `o` alias in some constant somewhere (in another PR).

Reply:
It could be easily retried by `$queryBuilder->getRootAliases()[0]`. Should I replace it here?

Reply:
don't IMO it's fine like this. It's just that we have a lot of these `'o'` alias hard-coded in the bridge.